### ISSUE TITLE
Git Bash significantly slower after upgrade to 2.41.0

### ISSUE THREAD

**Issue by sherman89** [git-for-windows/git#4459]
- Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

`64-bit version of Git 2.41.0.windows.1`

**I already downgraded to 2.40.1, apologies.**

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

Windows 11

```
Microsoft Windows [Version 10.0.22621.1778]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
Editor Option: Notepad++
Custom Editor Path: 
Default Branch Option:  
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

No

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git Bash on Windows Terminal

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

None, just upgraded to 2.41.0.

 - What did you expect to occur after running these commands?

Same performance as before.

 - What actually happened instead?

Everything is slower, cd to directory, git pull, git log, literally everything is significantly slower in Git Bash.

Every upgrade I did before throughout the years has been fine, except this one.

Downgrading to 2.40.1 fixed the issue.

**Comment by quyenlv-unicloud** [comment]
+1, it take 30s to start in my computer. Promon show new version just waste 10s for query "C:\Program Files\Git\usr\bin\DNSAPI.dll" and got "NAME NOT FOUND",

while old query "C:\Windows\System32\dnsapi.dll" and immediately got "SUCCESS"

**Comment by mormegil-cz** [comment]
Same here, the problem is with the bundled GNU utils. Git itself is fine and some built-in functionalities work normally, e.g. `git status`. However, any GNU tool takes several seconds to run. Just running e.g. `"c:\Program Files\Git\usr\bin\arch.exe"` takes ~20 seconds. So any git command running a `bash` subshell or `less` or anything takes very long time to execute.

**Comment by bricss** [comment]
Surprisingly, I cannot confirm any of this ❌ 
The latest Git version is blazingly fast ⚡  on both of my 2️⃣ machines with the latest Win11 x64 (OS build: `22621.1778`)

**Comment by Osyotr** [comment]
It's slow on Windows 10 but OK on Windows 11.

**Comment by mormegil-cz** [comment]
For reasons unknown to me, when running even just yes.exe, it seems to run LDAP queries against the Windows domain controller. Which seem to be correlated to the startup delay (but dunno, it _might_ be coincidence).

![Wireshark capture of the LDAP traffic](https://github.com/git-for-windows/git/assets/432985/fb7748c9-847d-4824-955c-2bd2bbf4a2ee)

![Thread stack captured by WPA](https://github.com/git-for-windows/git/assets/432985/d4e87182-9b68-4a23-ae1b-3b7e715d0b51)
(I don’t have symbols for msys-2.0.dll.)

Which… might be an upstream issue in msys2 maybe? (And it might also be caused by our DC being misconfigured or something, but still… I don’t think yes.exe needs to depend on proper LDAP configuration…)

**Comment by razvan-pricope** [comment]
We experience the same issue in our organization. The common thing for all of the employees is that the computers experiencing the slowdown are domain joined. Maybe this is the reason it's not 100% reproducible.

**Comment by PhilipOakley** [comment]
Domain controller delay problems show up moderately often. Not easy to resolve. Worth searching the archives (closed issues ;-)

**Comment by mormegil-cz** [comment]
OK, that did not really work, since the created passwd file had home directories pointing to `/home/username` which is wrong (unix-y), it needs to point to `/c/Users/username` so that ssh works; I had to modify the path manually. (Without it, `git fetch` did not find `known_hosts` and even when I confirmed the fingerprint, it failed with “Failed to add the host to the list of known hosts”.)

**Comment by MikeCheel** [comment]
git branch command takes excruciatingly long (i.e. minutes) to return. Downgrading to previous version of git solves the problem.

**Comment by rglidden** [comment]
Having similar performance issues on Windows 11 under PowerShell (not using git bash).  Some commands like `git status` are fine, but other commands like `git branch` and `git fetch` take forever. Large internal repos hosted in our on-prem Azure DevOps are excruciatingly slow. Smaller Github-hosted repos are a little better, but still noticeably slower. Machine is domain-joined. Have not yet tested on a non-domain joined machine.

Reverting to 2.40.1 restored all performance back to acceptable levels.

**Comment by geniuscodemonkey** [comment]
I'm seeing a git pull taking over 4 minutes. Half the time GIT will also crash... (I can't upload an image due to work security policy)
I've had to downgrade git to 2.39

**Comment by MikeCheel** [comment]
This problem is hapening in Windows 10 using powershell as well. I can further confirm that, as mentioned in this discussion previously, the gnu utils seem to be the culprit. Running the command "c:\Program Files\Git\usr\bin\arch.exe" to around 30 seconds to return. Running under admin prompt makes no difference.

**Comment by dscho** [comment]
I cannot reproduce here. For those who can, do you see the same issue with regular [Cygwin](https://cygwin.com/) (on which Git's Bash is based)?

**Comment by ianichitei** [comment]
> I cannot reproduce here. For those who can, do you see the same issue with regular [Cygwin](https://cygwin.com/) (on which Git's Bash is based)?

Using [cygwin 3.4.6.1](http://mirrors.kernel.org/sourceware/cygwin/x86_64/release/cygwin/cygwin-3.4.6-1-src.tar.xz) and simply running any `.exe` from that archive takes more than 20 seconds:

```console
$ Measure-Command { .\ldd.exe --help }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 22
Milliseconds      : 552
Ticks             : 225529747
TotalDays         : 0.000261029799768519
TotalHours        : 0.00626471519444444
TotalMinutes      : 0.375882911666667
TotalSeconds      : 22.5529747
TotalMilliseconds : 22552.9747
$ .\ldd.exe --version
ldd (cygwin) 3.4.6
Print shared library dependencies
Copyright (C) 2009 - 2023 Chris Faylor
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by ianichitei** [comment]
I tried with 3.4.7 and the issue is not fixed. I'll try to report this directly to the Cygwin devs.

**Comment by ianichitei** [comment]
Here is the [Cygwin bug report](https://sourceware.org/pipermail/cygwin/2023-June/253895.html).

**Comment by ianichitei** [comment]
Based on the responses from the Cygwin mailing list I tried the second workaround from [here](https://cygwin.com/faq.html#faq.using.startup-slow) and it solved the problem for me: Cygwin Terminal now starts instantly. I don't know how one could do something similar for the Cygwin bundled with Git for Windows.

**Comment by dscho** [comment]
> I tried the second workaround from [here](https://cygwin.com/faq.html#faq.using.startup-slow) and it solved the problem for me

You mean this?

Run these commands in a Cygwin terminal with write access to `/etc`:

```
getent passwd $(id -u) > /etc/passwd
getent group $(id -G) > /etc/group
```

Also, set `/etc/nsswitch.conf` as follows:

```
passwd: files db
group:  files db
```

If so, we already have the latter (even more than that: we comment out the `db` part in the `group` line). But the former only works in single-user setups, which Git for Windows does not want to be limited to.

> It is a bit weird that it started to happen to different people from different organizations after updating to 2.41.0.

Indeed. The only thing we did that is notable was to upgrade from MSYS2 runtime v3.3.* to v3.4.*. But there were so many changes between those major versions that I would suspect some regression to have crept in.

**Comment by ianichitei** [comment]
> You mean this? [...] If so, we already have the latter (even more than that: we comment out the db part in the group line). But the former only works in single-user setups, which Git for Windows does not want to be limited to.

Yes, exactly those, but if that is already done in case of Git it probably means that there's another cause for the problem, and the slowdown I experienced with Cygwin has nothing to do with the one experienced with Git for Windows. 

> Indeed. The only thing we did that is notable was to upgrade from MSYS2 runtime v3.3.* to v3.4.*. But there were so many changes between those major versions that I would suspect some regression to have crept in.

I don't really know how to further reply to the Cygwin issue with regards to this, but maybe someone on your side could add some context to that bug report?

**Comment by jankap** [comment]
I have the same issues at my corporate workstation. As long as I'm connected to our VPN (which is required for most workflows), Git is painfully slow. `git config --global --list` is slow, `git status` is slow, `git commit --amend` (which calls VSCode) is slow. Disabling the VPN and everything is snappy again.

How can I help to debug that?

**Comment by dscho** [comment]
> looks like I'm on msys, not cygwin:

You're actually on MSYS2, and both MSys and MSYS2 are forks of Cygwin, so for the purposes of this ticket it is irrelevant on which system you are.

Disabling the VPN most likely cuts off the connection to the domain controller and the slow calls are simply skipped, which is consistent with what we knew already.

What we still don't know is which parts of Cygwin's code is executed in the slow path, and most crucially: _why_ it is executed when we do not really need all that information that the domain controller is asked to report.

What _would_ be helpful, therefore, would be a thorough debugging session that most likely requires instrumenting (and [building](https://github.com/git-for-windows/git/wiki/Building-msys2-runtime)) the MSYS2 runtime and repeatedly running simple commands both with the VPN enabled and disabled to identify above-mentioned code paths.

**Comment by rglidden** [comment]
I spent some time looking through diffs between MSYS2 3.3 and MSYS2 3.4, and I suspect the issue is related to https://github.com/git-for-windows/msys2-runtime/commit/a5bcfe616c7e8f78f464bf045595d8213244876a which removes a short-circuit for the current user and instead reaches out to AD for user information.  This might also explain why having the "db" portion of the groups line in `/etc/nsswitch.conf` doesn't prevent this problem from occurring.

I don't really have the experience necessary to figure out how to build MSYS2 and get Git to use it to test it out quickly. I may use it as a learning experience, but just wanted to point this out in case anyone more experienced with building MSYS2 and git can give it a quick test, or provide me with some test binaries to try.

**Comment by dscho** [comment]
> I suspect the issue is related to [git-for-windows/msys2-runtime@a5bcfe6](https://github.com/git-for-windows/msys2-runtime/commit/a5bcfe616c7e8f78f464bf045595d8213244876a) which removes a short-circuit for the current user and instead reaches out to AD for user information.

Great work, @rglidden, that indeed looks like a plausible culprit!

> I don't really have the experience necessary to figure out how to build MSYS2 and get Git to use it to test it out quickly.

I'm happy to report that it's not all that difficult to get started, in particular because we have tons of automation by now.

All you've got to do is to fork & clone https://github.com/git-for-windows/msys2-runtime/, `git revert` that commit locally and then open a PR. This will trigger an automated build in the "Checks" tab, called "build on: pull_request" ([example](https://github.com/git-for-windows/msys2-runtime/pull/56/checks)). Once the build is done, you can click on that word "build", which is a link and will lead you to the summary where you can see the `install` build artifact ([example](https://github.com/git-for-windows/msys2-runtime/actions/runs/5666805417)). This artifact is a `.zip` file that contains, among other files, `usr\bin\msys-2.0.dll`, which you can drop into, say, an unpacked PortableGit (e.g. [v2.42.0](https://github.com/git-for-windows/git/releases/download/v2.42.0.windows.1/PortableGit-2.42.0-64-bit.7z.exe)), overwriting the existing `usr\bin\msys-2.0.dll`. Then you can call `git-bash.exe` from that PortableGit's top-level directory to test.

If this is indeed the culprit, I'll help you introduce a new setting for the `nsswitch.conf` that we will use in Git for Windows from now on, to enable the fast path again.

**Comment by rglidden** [comment]
Thanks for the detailed instructions @dscho. I was trying to figure out how to build things locally. I didn't realize I could leverage your build infrastructure by submitting a PR. That will make things a LOT easier!

I'll see if I can find some time to try it in the next couple days and will let you know what I find.

**Comment by dscho** [comment]
Excellent!

**Comment by rglidden** [comment]
Looks like my hunch was correct!

1. Verified that PortableGit-2.42.0 still suffers the performance problem on my machine. ~30 seconds to launch git bash or do most git commands.
2. Copied the `msys-2.0.dll` from https://github.com/git-for-windows/msys2-runtime/pull/57 into `usr/bin/msys-2.0.dll` in my PortableGit folder.
3. Problem is resolved. Git bash loads nearly instantly, and all git commands are now responsive.

**Comment by dscho** [comment]
@rglidden awesome work, and I am so glad that your hunch was correct and now you have proof!

So, let me walk you through the process how to add a keyword to `/etc/nsswitch.conf` and how to let that keyword toggle how the code behaves.

The central file to do this, the file [where that `db` keyword is parsed](https://github.com/cygwin/cygwin/blob/cygwin-3_4-branch/winsup/cygwin/uinfo.cc#L635), is `winsup/cygwin/uinfo.cpp`. You can see that the `NSS_SRC_DB` flag is set when `db` is matched. Whether this `NSS_SRC_DB` flag is set or not happens to be queried exclusively via [`nss_pwd_db()`](https://github.com/cygwin/cygwin/blob/cygwin-3_4-branch/winsup/cygwin/local_includes/cygheap.h#L452), most notably in [`internal_getpwsid()`](https://github.com/cygwin/cygwin/blob/208c3f68a30a7ab6cc53f30f8a119694ed29c866/winsup/cygwin/passwd.cc#L110) and in [`internal_getpwnam()`](https://github.com/cygwin/cygwin/blob/208c3f68a30a7ab6cc53f30f8a119694ed29c866/winsup/cygwin/passwd.cc#L149).

You will probably realize that despite `passwd: files db` suggesting an order, suggestive of a precedence of "files" over "db", the precedence is the same if you write `passwd: db files` (still "files" over "db") because the values are parsed into the bit field `pwd_src`, and the code will always first query `nss_pwd_files()` and only when that did not find anything the code goes on to the `if (nss_pwd_db())` clause.

This is good! Because it means that we do not have to deal with precedence other than the hard-coded one. We can even add a keyword, say, `db-quick-and-dirty` (I did not really find a word reflecting what we want to do here, so I went with my old band's name) that includes the meaning of `db`, by letting it set `NSS_SRC_DB` _and_ another flag, say, `NSS_SRC_DERIVE_GROUP_FROM_CURRENT_USER_TOKEN` that we can then use in `fetch_account_from_windows()` to reinstate the fast code path.

But then, we probably want to default to the fast 'n furious code path in the MSYS2 runtime, and only do the correct-yet-slow work if asked to do so, [as we all do](https://en.wikipedia.org/wiki/Thinking,_Fast_and_Slow)? And while I talked about `passwd` all the time, I now realize that the `group: files db` line is a much more appropriate target for this here discussion: after all, the problematic slow-down happens because we want to fetch the current user's default _group_, and the fast path simply gets it from the current user token, which is sometimes incorrect.

So here is what I'd suggest to do:

- Amend the commit https://github.com/git-for-windows/msys2-runtime/commit/a51cbf8220a08668e5d277af9c9703be2b7b24ab with the following:
- Add a new `NSS_SRC_ACCURATE = 4` constant next to [`NSS_SRC_FILES` and `NSS_SRC_DB`](https://github.com/cygwin/cygwin/blob/208c3f68a30a7ab6cc53f30f8a119694ed29c866/winsup/cygwin/include/sys/cygwin.h#L212-L217).
- Introduce a new `group` keyword in the `/etc/nsswitch.conf` file, say, `db-accurate`, and parse it like `db`, but assign `*src |= NSS_SRC_DB | NSS_SRC_ACCURATE`. Maybe print a warning if the keyword is used in a `passwd` line because it won't have any effect. If you prefer, you can also name it `accurate` and only assign `*src |= NSS_SRC_ACCURATE`, but that has the downside that a `group: files accurate` line will be valid but the `accurate` won't do anything, which might surprise some.
- Add a new `nss_grp_accurate()` method [next to `nss_grp_files()` and `nss_grp_db()`](https://github.com/cygwin/cygwin/blob/208c3f68a30a7ab6cc53f30f8a119694ed29c866/winsup/cygwin/local_includes/cygheap.h#L454-L455).
- Rename the [`is_current_user`](https://github.com/git-for-windows/msys2-runtime/commit/a51cbf8220a08668e5d277af9c9703be2b7b24ab#diff-1ffeda03bc188fa732454f52c4932977bc8233d9db4da19ab5acb0c58c7320ccR1909) variable to something more meaningful, e.g. `get_default_group_from_current_user_token`.
- Augment the [`sid == cygheap->user.sid ()` condition guarding the fast path](https://github.com/git-for-windows/msys2-runtime/commit/a51cbf8220a08668e5d277af9c9703be2b7b24ab#diff-1ffeda03bc188fa732454f52c4932977bc8233d9db4da19ab5acb0c58c7320ccR2379) using the newly-introduced `nss_grp_accurate()` method, i.e. `if (!cygheap->pg.nss_grp_accurate () && sid == cygheap->user.sid ())`.
- Document the new keyword [in the appropriate section](https://github.com/cygwin/cygwin/blob/27376c60a9b870e7ba456551165897a7408b5886/winsup/doc/ntsec.xml#L925-L987), taking care to mention that it only has an effect on the `group` line. Use the information from the commit message of https://github.com/git-for-windows/msys2-runtime/commit/a5bcfe616c7e8f78f464bf045595d8213244876a to illustrate why users may want to use that flag.
- Write a kick-ass commit message to motivate all of this, mentioning that "This fixes https://github.com/git-for-windows/git/issues/4459".

Questions? 😁

**Comment by clach04** [comment]
This is a me-too comment :-D

1. I have a machine **not** on a domain, with 2.31.0 which is fine.
2. I have a machine **on**  a domain, tested with 2.41.0.3-64-bit and latest (as of 2023-08-25) Git-2.42.0-64-bit - which is slow (for diff).

Originally I thought https://github.com/git-for-windows/git/issues/4574 was different to this issue #4459 but I'm now suspecting they are the same based on the pager test suggested in that issue.

**Comment by rglidden** [comment]
Thanks again for all the extremely detailed instructions @dscho! This has been a very helpful and welcoming introduction to submitting changes to git for Windows. 😄

I've updated https://github.com/git-for-windows/msys2-runtime/pull/57 with the changes you suggested. Tested locally and it resolves the performance issues, while preserving the more "correct" behaviour by adding the `group: db-accurate` option.

Please review when you can at let me know if anything needs adjusting.

**Comment by dscho** [comment]
@rglidden thank you so much! I am in the process of integrating this into Git for Windows (will probably have to take a break and finish it tomorrow).

**Comment by sherman89** [comment]
@rglidden Thank you very much! :)
